server-spi-private/src/main/java/org/keycloak/userprofile/DefaultUserProfile.java

```diff
@@ -0,0 +1,147 @@
+    private final Attributes attributes;
+    private boolean validated;
+    private UserModel user;
+
+    public DefaultUserProfile(Attributes attributes, Function<Attributes, UserModel> userCreator, UserModel user) {
+        this.userSupplier = userCreator;
+        this.attributes = attributes;
+        this.user = user;
+    }
+
+    @Override
+    public void validate() {
+        ValidationException validationException = new ValidationException();
+
+        for (String attributeName : attributes.nameSet()) {
+            this.attributes.validate(attributeName,
+                    (attribute, message) -> validationException.addError(new ValidationException.Error(attribute, message)));
+        }
+
+        if (validationException.hasError()) {
+            throw validationException;
+        }
+
+        validated = true;
+    }
+
+    @Override
+    public UserModel create() throws ValidationException {
+        if (user != null) {
+            throw new RuntimeException("User already created");
+        }
+
+        if (!validated) {
+            validate();
+        }
+
+        user = userSupplier.apply(this.attributes);
+
+        return updateInternal(user, false);
+    }
+
+    @Override
+    public void update(boolean removeAttributes, BiConsumer<String, UserModel>... changeListener) {
+        if (!validated) {
+            validate();
+        }
+
+        updateInternal(user, removeAttributes, changeListener);
+    }
+
+    private UserModel updateInternal(UserModel user, boolean removeAttributes, BiConsumer<String, UserModel>... changeListener) {
+        if (user == null) {
+            throw new RuntimeException("No user model provided for persisting changes");
+        }
+
+        try {
+            for (Map.Entry<String, List<String>> attribute : attributes.attributeSet()) {
+                String name = attribute.getKey();
+
+                if (attributes.isReadOnly(name)) {
+                    continue;
+                }
+
+                List<String> currentValue = user.getAttributeStream(name).filter(Objects::nonNull).collect(Collectors.toList());
+                List<String> updatedValue = attribute.getValue().stream().filter(Objects::nonNull).collect(Collectors.toList());
+
+                if (currentValue.size() != updatedValue.size() || !currentValue.containsAll(updatedValue)) {
+                    user.setAttribute(name, updatedValue);
+                    for (BiConsumer<String, UserModel> listener : changeListener) {
+                        listener.accept(name, user);
+                    }
+                }
+            }
+
+            // this is a workaround for supporting contexts where the decision to whether attributes should be removed depends on
+            // specific aspect. For instance, old account should never remove attributes, the admin rest api should only remove if
+            // the attribute map was sent.
+            if (removeAttributes) {
```

Comment:
Is this really correct implementation for new functionality? 
Legacy code removed all attributes which did not have new value (except read only ones) - this should stay in legacy provider to be backward compatible. 
But new implementations should touch only attributes which are present in UserProfile selected for current action, which means it should only remove attribute if it is configured for current UserProfile but new value is empty. It should never touch/remove attribute which is not in currently selected UserProfile.

Reply:
You asked that before. This is due to how we are managing attributes at some places, where the decision to whether or not attributes should be removed depends on a specific logic. 

See the `UsersResource` changes. This is the only place we use that. This is not really related to the user profile itself but how we do certain things. There if attributes is null we don't remove them, if empty we remove them. Differently than other places where we always expect the full view of attributes.

Ideally, we should support PATCH in some endpoints. As per our rest guidelines.

Reply:
I know we discussed `removeAttributes` parameter already, why it exists, and I hope I understand why it is here :-D But still I'm not sure if it is used correctly, why are all attributes which do not have new value in `attributes` instance removed based on it. For me it looks like UserProfile configuration selected for the current action is not reflected here, as we should remove only attributes which are part of the user profile and do not have new value, but we should never touch others. Maybe I misunderstand something, I'll try to do some tests once your PR is merged to see how exactly this code works

Reply:
The `UserProfile` only processes the attributes that are defined in the configuration.

The reasoning behind that `removeAttributes` flag is more related to whether you should remove attributes or not. As an example, in the old account console we never remove attributes. In the admin rest api, we never remove attributes if the representation is null.

This is just to align with the "oddness" we have in some places. I did not manage to avoid that flag but I would love to.